src/app-layer-ssl.c

```diff
@@ -1171,6 +1197,106 @@ static inline int TLSDecodeHSHelloExtensionEllipticCurvePF(SSLState *ssl_state,
     return -1;
 }
 
+static inline int TLSDecodeHSHelloExtensionSigAlgorithms(
+        SSLState *ssl_state, const uint8_t *const initial_input, const uint32_t input_len)
+{
+    const uint8_t *input = initial_input;
+
+    /* Empty extension */
+    if (input_len == 0)
+        return 0;
+
+    if (!(HAS_SPACE(2)))
+        goto invalid_length;
+
+    uint16_t sigalgo_len = (uint16_t)(*input << 8) | *(input + 1);
+    input += 2;
+
+    /* Signature algorithms length should always be divisible by 2 */
+    if ((sigalgo_len % 2) != 0) {
+        goto invalid_length;
+    }
+
+    if (!(HAS_SPACE(sigalgo_len)))
+        goto invalid_length;
+
+    if (ssl_state->curr_connp->ja4 != NULL &&
+            ssl_state->current_flags & SSL_AL_FLAG_STATE_CLIENT_HELLO) {
+        uint16_t sigalgo_processed_len = 0;
+        while (sigalgo_processed_len < sigalgo_len) {
+            uint16_t sigalgo = (uint16_t)(*input << 8) | *(input + 1);
+            input += 2;
+            sigalgo_processed_len += 2;
+
+            SCJA4AddSigAlgo(ssl_state->curr_connp->ja4, sigalgo);
+        }
+    } else {
+        /* Skip signature algorithms */
+        input += sigalgo_len;
+    }
+
+    return (input - initial_input);
+
+invalid_length:
+    SCLogDebug("Signature algorithm list invalid length");
+    SSLSetEvent(ssl_state, TLS_DECODER_EVENT_HANDSHAKE_INVALID_LENGTH);
+
+    return -1;
+}
+
+static inline int TLSDecodeHSHelloExtensionALPN(
+        SSLState *ssl_state, const uint8_t *const initial_input, const uint32_t input_len)
+{
+    const uint8_t *input = initial_input;
+
+    /* Empty extension */
+    if (input_len == 0)
+        return 0;
+
+    if (!(HAS_SPACE(2)))
+        goto invalid_length;
+
+    uint16_t alpn_len = (uint16_t)(*input << 8) | *(input + 1);
+    input += 2;
+
+    if (!(HAS_SPACE(alpn_len)))
+        goto invalid_length;
+
+    if (ssl_state->curr_connp->ja4 != NULL &&
+            ssl_state->current_flags & SSL_AL_FLAG_STATE_CLIENT_HELLO) {
+        /* We use 32 bits here to avoid potentially overflowing a value that
+           needs to be compared to an unsigned 16-bit value. */
+        uint32_t alpn_processed_len = 0;
+        while (alpn_processed_len < alpn_len) {
+            uint8_t protolen = *input;
+            input += 1;
+            alpn_processed_len += 1;
+
+            if (!(HAS_SPACE(protolen)))
+                goto invalid_length;
+
+            /* we only want the first value for JA4 */
+            if (alpn_processed_len == 1) {
+                SCJA4SetALPN(ssl_state->curr_connp->ja4, (const char *)input, protolen);
+            }
+
+            alpn_processed_len += protolen;
```

Comment:
Still sad that this can overflow

Reply:
Why?
* `alpn_len` is a 16-bit unsigned value.
* `alpn_processed_len` is a 32-bit unsigned value.
* `protolen` is a 8-bit unsigned value, i.e. `alpn_processed_len` can only be incremented each iteration by at most 255.
* So the only way that this can overflow is that `alpn_processed_len` + `protolen` exceeds the range of a 32-bit unsigned value -- but that cannot happen since in that case `alpn_processed_len` would have had to have exceeded any value of `alpn_len` many iterations ago, since we only can increment `alpn_processed_len` by 255 at most. But that would have already been caught by the `while (alpn_processed_len < alpn_len)` comparison at the largest 16-bit unsigned value at the latest, so we couldn't even get into that region.

What am I missing?

Reply:
Oh yes,

> alpn_processed_len is a 32-bit unsigned value 

This is good

But still, I can have input going further than `alpn_len` if the last `protolen` is too big, and so we may parse bad wrongly next extensions...

Reply:
Ah, so your comment was not about an integer _type_ overflow (that might cause a loop), but rather a semantic error to be able to exceed `alpn_len`...

Reply:
My comment was more generically speaking about a missing check and all that could go wrong because of it ;-)

Reply:
So I guess it would be safer to do something like that, to recheck whether it's OK to read from the buffer and increment `input`?
```diff
diff --git a/src/app-layer-ssl.c b/src/app-layer-ssl.c
index 361a81696..16cc2bc66 100644
--- a/src/app-layer-ssl.c
+++ b/src/app-layer-ssl.c
@@ -1275,7 +1275,14 @@ static inline int TLSDecodeHSHelloExtensionALPN(
             if (!(HAS_SPACE(protolen)))
                 goto invalid_length;
 
-            /* we only want the first value for JA4 */
+            /* Check if this length would exceed the overall ALPN length;
+               if so, skip and continue */
+            if (alpn_processed_len + protolen > ((uint32_t) alpn_len)) {
+                input += alpn_len - alpn_processed_len;
+                break;
+            }
+
+            /* Only record the first value for JA4 */
             if (alpn_processed_len == 1) {
                 SCJA4SetALPN(ssl_state->curr_connp->ja4, (const char *)input, protolen);
             }

```